Source/WebCore/loader/FrameLoader.cpp

```diff
@@ -4742,17 +4746,40 @@ void FrameLoader::updateNavigationAPIEntries()
     RefPtr page = m_frame->page();
     if (!page)
         return;
-    if (RefPtr currentItem = m_frame->history().currentItem()) {
-        Vector<Ref<HistoryItem>> historyItems;
-        if (currentItem == page->backForward().currentItem()) {
-            for (int index = -1 * static_cast<int>(page->backForward().backCount()); index <= static_cast<int>(page->backForward().forwardCount()); index++) {
-                if (RefPtr item = page->backForward().itemAtIndex(index))
-                    historyItems.append(item.releaseNonNull());
-            }
-        } else
-            historyItems.append(*currentItem);
-        domWindow->protectedNavigation()->initializeEntries(*currentItem, historyItems);
+    RefPtr currentItem = m_frame->history().currentItem();
+    if (!currentItem)
+        return;
+
+    // https://html.spec.whatwg.org/multipage/browsing-the-web.html#getting-session-history-entries-for-the-navigation-api
+    Vector<Ref<HistoryItem>> entriesForNavigationAPI;
+    entriesForNavigationAPI.append(*currentItem);
+
+    auto rawEntries = page->backForward().allItems();
+    auto startingIndex = rawEntries.find(*currentItem);
+    if (startingIndex != notFound) {
+        Ref startingOrigin = SecurityOrigin::create(rawEntries[startingIndex]->url());
+
+        int64_t i = startingIndex - 1;
+        while (i >= 0) {
+            Ref item = rawEntries[i];
+            if (!SecurityOrigin::create(item->url())->isSameOriginAs(startingOrigin))
+                break;
+
+            entriesForNavigationAPI.insert(0, WTFMove(item));
+            i -= 1;
```

Comment:
This is what -- is for.  Also, stylistically this would be a little nicer as a for loop.  The variable i doesn't even need to be the same variable as the next loop because its value is reset.